### PR TITLE
VAA-3 add CrewMemberDao

### PR DESCRIPTION
**Pull request by vitaly-lab**

### CODE REVIEW

File: src/main/java/com/airlines/dao/CrewMemberDao.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.airlines.dao;
+
+import com.airlines.entity.CrewMember;
+
+public interface CrewMemberDao {
+
+    void save(CrewMember account);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/CrewMemberDao.java]: why object type is CrewMember and parameter type is "Account"?

File: src/main/java/com/airlines/dao/CrewMemberDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.Date;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class CrewMemberDaoImpl implements CrewMemberDao {
+
+    private final static String INSERT_CREWMEMBER_SQL = "INSERT INTO crew_members(first_name, last_name, position, birthday, citizenship) VALUES (?, ?, ?, ?, ?);";
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/CrewMemberDaoImpl.java]: INSERT_CREWMEMBER_SQL -> INSERT_CREW_MEMBER_SQL

and so on for constants

File: src/main/java/com/airlines/dao/CrewMemberDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    private final static String SELECT_CREWMEMBER_BY_ID_SQL = "SELECT * FROM crew_members WHERE id = ?;";
+    private final static String UPDATE_CREWMEMBER_SQL = "UPDATE crew_members SET first_name =?, last_name = ?, position = ?, birthday = ?, citizenship = ? WHERE id = ?;";
+    private DataSource dataSource;
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/CrewMemberDaoImpl.java:21]: public static final 
public static final 

private static final ...
private static final ...
private static final ...
`                                                             <---empty line`
private
private

Class fields are grouped by type - constants (static final first ), by modificator - public, then private and so on:
public
protected
[default]
private

...static final

...final

...private

File: src/main/java/com/airlines/dao/CrewMemberDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        } else {
+            throw new DaoOperationException("Can not obtain an account ID");
+        }
+    }
+
+    public CrewMember findOne(int id) {
+        try (Connection connection = dataSource.getConnection()) {
+            return findCrewMemberById(id, connection);
+        } catch (SQLException e) {
+            throw new DaoOperationException(String.format("Cannot find Account by id = %d", id), e);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/CrewMemberDaoImpl.java]: Account -> crew member

And please use one approach for the symmetry principle - or "id" or "ID"

File: src/main/java/com/airlines/dao/CrewMemberDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                .withPosition(Position.BOARD_CONDUCTOR)
+                .withPosition(Position.FIRST_PILOT)
+                .withPosition(Position.STEWARDESS)
+                .withPosition(Position.FIRST_PILOT)
+                .withCitizenship(Citizenship.AUSTRIA)
+                .withCitizenship(Citizenship.AUSTRALIA)
+                .withCitizenship(Citizenship.GERMANY)
+                .withCitizenship(Citizenship.NEW_ZEALAND)
+                .withCitizenship(Citizenship.UK)
+                .withCitizenship(Citizenship.USA)
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/CrewMemberDaoImpl.java]: Imagine you have Cat class:

```
@Builder
class Cat {
   private int age;
   private String name;
}
```

this method is wehere you get information (row / resultset) from database and parse into your entity.
so you have Cat("murka", 25) from database: and doing next:

Cat.builder()
   .name("tuzik")
   .name("murka")
   .name("furryik")
    .age(10)
    .age(5)
    .age(10)
    .build();
    
    but all you need just:
     Cat.build()
        .name(resultSet.getString("name"))
        .age(resultSet.getString("age))
        .build();
        
        Enum is just ordinary string (varchar) in database. All you need here - is just get string from database and parse it into enum:
         .withPosition(Position.valueOf(resultSet.getString("position"))
          smth like that

File: src/main/java/com/airlines/dao/CrewMemberDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            PreparedStatement selectByIdStatement = connection.prepareStatement(SELECT_CREWMEMBER_BY_ID_SQL);
+            selectByIdStatement.setLong(1, id);
+            return selectByIdStatement;
+        } catch (SQLException e) {
+            throw new DaoOperationException("Cannot prepare statement to select account by id", e);
+        }
+    }
+
+    private CrewMember parseRow(ResultSet resultSet) throws SQLException {
+        CrewMember crewMember = CrewMember.builder()
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/CrewMemberDaoImpl.java]: you don'n need this variable, you can write just:
private CrewMember parseRow(ResultSet resultSet) throws SQLException {
        return CrewMember.builder()...

File: src/main/java/com/airlines/dao/CrewMemberDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                .build();
+
+        return crewMember;
+    }
+
+    @Override
+    public void update(CrewMember crewMember) {
+        try (Connection connection = dataSource.getConnection()) {
+            PreparedStatement updateStatement = prepareUpdateStatement(crewMember, connection);
+            executeUpdate(updateStatement, "Account was not updated");
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/CrewMemberDaoImpl.java]: CrewMember was not updated

**Review by joseph-dredd — CHANGES_REQUESTED** [review]

File: src/main/java/com/airlines/dao/CrewMemberDao.java (hunk 1 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/CrewMemberDao.java]: fixed in 45b9bb5fd29cdf8936e2dab8daa2a25ad85a7b59

File: src/main/java/com/airlines/dao/CrewMemberDaoImpl.java (hunk 2 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/CrewMemberDaoImpl.java]: fixed in 45b9bb5fd29cdf8936e2dab8daa2a25ad85a7b59

File: src/main/java/com/airlines/dao/CrewMemberDaoImpl.java (hunk 3 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/CrewMemberDaoImpl.java:21]: fixed in ce44c93c4a6ed5eccdb3296f2989f672155d8cbd

**Review by joseph-dredd — APPROVED** [review]

**Pull request merged**